XPointMLTest.py

```diff
@@ -857,53 +948,116 @@ def main():
+
+    if use_amp:
+        if args.amp_dtype == 'bfloat16' and not torch.cuda.is_bf16_supported():
+            print("Warning: bfloat16 not supported on this GPU. Falling back to float16.")
+        print(f"Using Automatic Mixed Precision (AMP) with dtype: {amp_dtype}")
 
     checkpoint_dir = "checkpoints"
     os.makedirs(checkpoint_dir, exist_ok=True)
     latest_checkpoint_path = os.path.join(checkpoint_dir, "xpoint_model_latest.pt")
     start_epoch = 0
     train_loss = []
     val_loss = []
+    best_val_loss = float('inf')
 
     if os.path.exists(latest_checkpoint_path) and not args.smoke_test:
         model, optimizer, start_epoch, train_loss, val_loss = load_model_checkpoint(
             model, optimizer, latest_checkpoint_path
         )
         print(f"Resuming training from epoch {start_epoch+1}")
+        print(f"Best validation loss so far: {best_val_loss:.6f}")
     else:
         print("Starting training from scratch")
 
     t2 = timer()
     print("time (s) to prepare model: " + str(t2-t1))
 
+    # Early stopping setup
+    patience_counter = 0
+    
     num_epochs = args.epochs
     for epoch in range(start_epoch, num_epochs):
-        train_loss.append(train_one_epoch(model, train_loader, criterion, optimizer, device))
-        val_loss.append(validate_one_epoch(model, val_loader, criterion, device))
-        print(f"[Epoch {epoch+1}/{num_epochs}]  TrainLoss={train_loss[-1]} ValLoss={val_loss[-1]}")
+        train_loss_epoch = train_one_epoch(model, train_loader, criterion, optimizer, device, scaler, use_amp, amp_dtype)
+        val_loss_epoch = validate_one_epoch(model, val_loader, criterion, device, use_amp, amp_dtype)
+        
+        train_loss.append(train_loss_epoch)
+        val_loss.append(val_loss_epoch)
         
-        # Save model checkpoint after each epoch
-        if epoch % args.checkPointFrequency == 0:
-          save_model_checkpoint(model, optimizer, train_loss, val_loss, epoch+1, checkpoint_dir)
+        current_lr = optimizer.param_groups[0]['lr']
+        print(f"[Epoch {epoch+1}/{num_epochs}] LR={current_lr:.2e} TrainLoss={train_loss[-1]:.6f} ValLoss={val_loss[-1]:.6f}")
+        
+        # Learning rate scheduling
+        scheduler.step()
+        
+        # Check for improvement
+        if val_loss[-1] < best_val_loss:
+            best_val_loss = val_loss[-1]
+            patience_counter = 0
+            print(f"   New best validation loss: {best_val_loss:.6f}")
+            # Save best model
+            torch.save(model.state_dict(), os.path.join(checkpoint_dir, "best_model.pt"))
+        else:
+            patience_counter += 1
+            
+        # Save checkpoint periodically
+        if (epoch+1) % args.checkPointFrequency == 0:
+            save_model_checkpoint(model, optimizer, train_loss, val_loss, epoch+1, checkpoint_dir, scaler, best_val_loss)
+        
+        # Early stopping
+        if patience_counter >= args.patience:
```

Comment:
This will stop training if the training/value loss isn't decreasing sufficiently?  If so, can you expand the doc string to indicate that.

Reply:
> This will stop training if the training/value loss isn't decreasing sufficiently
As discussed offline, yes.  'patience' is common a ML/AI training term/concept.